contracts/external/proxies/PayableCERC20.sol

```diff
@@ -0,0 +1,134 @@
+
+        controller = Controller(_controller);
+        marginPool = _marginPool;
+        cToken = CERC20Interface(_cToken);
+        underlying = ERC20Interface(cToken.underlying());
+        weth = WETH9(_weth);
+
+        ERC20Interface(address(weth)).safeApprove(_marginPool, uint256(-1));
+    }
+
+    /**
+     * @notice fallback function which disallows ETH to be sent to this contract without data except when unwrapping WETH
+     */
+    fallback() external payable {
+        require(msg.sender == address(weth), "PayableCERC20: Cannot receive ETH");
+    }
+
+    /**
+     * @notice execute a number of actions after minting some cTokens
+     * @dev a wrapper for the Controller operate function, to wrap uderlying to cToken before the excution.
+     * @param _actions array of actions arguments
+     * @param _amountUnderlying the amount of underlying to supply to Compound
+     */
+    function operate(
+        Actions.ActionArgs[] memory _actions,
+        address payable sendEthTo,
+        uint256 _amountUnderlying
+    ) external payable nonReentrant {
+        uint256 cTokenBalance;
+
+        // create WETH from ETH
+        if (msg.value != 0) {
+            weth.deposit{value: msg.value}();
+        }
+        if (_amountUnderlying > 0) {
+            underlying.safeTransferFrom(msg.sender, address(this), _amountUnderlying);
+            // mint cToken
+            underlying.safeIncreaseAllowance(address(cToken), _amountUnderlying);
+
+            require(cToken.mint(_amountUnderlying) == 0, "PayableCERC20: cToken mint failed");
+
+            cTokenBalance = cToken.balanceOf(address(this));
+            cToken.safeIncreaseAllowance(marginPool, cTokenBalance);
+        }
+
+        // verify sender
+        for (uint256 i = 0; i < _actions.length; i++) {
+            Actions.ActionArgs memory action = _actions[i];
+
+            // check that msg.sender is an owner or operator
+            if (action.owner != address(0)) {
+                require(
+                    (msg.sender == action.owner) || (controller.isOperator(action.owner, msg.sender)),
+                    "PayableCERC20: msg.sender is not owner or operator "
+                );
+            }
+
+            // overwrite the deposit amount by the exact amount minted
+            if (
+                action.actionType == Actions.ActionType.DepositCollateral &&
+                action.amount == 0 &&
+                action.asset == address(cToken)
+            ) {
+                // should we only update if the action.second address == address(this)?
+                _actions[i].amount = cTokenBalance;
+            }
+            // check if Call action exist, approve WETH to Callee
+            if (action.actionType == Actions.ActionType.Call) {
+                ERC20Interface(address(weth)).safeIncreaseAllowance(
+                    action.secondAddress,
+                    weth.balanceOf(address(this))
```

Comment:
```suggestion
                    msg.value
```
I think using msg.value is equivalent to `weth.balanceOf(address(this))` in all cases, and it saved us some gas